File: sol/632/632.cpp
#include <queue>

struct Solution {
    vector<int> smallestRange(vector<vector<int>>& nums) {
        int k = nums.size(), totCount = 0;
        vector<int> ptrs(k, 1), arrCount(k, 0);
        vector<pair<int, int>> h(k);
        for (int i = 0; i < k; ++i){
            h[i] = make_pair(-nums[i][0], i);
        }
        ranges::make_heap(h);
        deque<pair<int, int>> q;
        int wl = -1e5, wr = 1e5;
        while (!h.empty()){
            ranges::pop_heap(h);
            auto [v, i] = h.back();
            h.pop_back();
            if (ptrs[i] != nums[i].size()){
                h.emplace_back(-nums[i][ptrs[i]++], i);
                ranges::push_heap(h);
            }
            q.emplace_back(-v, i);
            if (++arrCount[i] == 1){
                ++totCount;
            }
            while (!q.empty() && totCount == k && arrCount[q.front().second] != 1){
                --arrCount[q.front().second];
                q.pop_front();
            }
            if (totCount == k && q.back().first - q.front().first < wr - wl){
                wl = q.front().first;
                wr = q.back().first;
            }
        }
        return {wl, wr};
    }
};

File: sol/632/632.py
import heapq as hq
from collections import deque

class Solution:
    def smallestRange(self, nums: List[List[int]]) -> List[int]:
        k = len(nums)
        ptrs = [1] * k
        arrCount = [0] * k
        totCount = 0
        h = [(nums[i][0], i) for i in range(k)]
        hq.heapify(h)
        q = deque()
        minWindow = [-10 ** 5, 10 ** 5]
        while len(h) != 0:
            v, arridx = hq.heappop(h)
            if ptrs[arridx] != len(nums[arridx]):
                hq.heappush(h, (nums[arridx][ptrs[arridx]], arridx))
                ptrs[arridx] += 1
            q.append((v, arridx))
            arrCount[arridx] += 1
            if arrCount[arridx] == 1:
                totCount += 1
            while len(q) != 0 and totCount == k and arrCount[q[0][1]] != 1:
                arrCount[q.popleft()[1]] -= 1
            if totCount == k \
                    and q[-1][0] - q[0][0] < minWindow[1] - minWindow[0]:
                minWindow = [q[0][0], q[-1][0]]
        return minWindow
